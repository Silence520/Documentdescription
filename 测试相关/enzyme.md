#enzyme

> Enzyme 是由 AirBnb 团队发布和维护的测试实用程序库, 它提供了一个更好的、高级的 API 来处理测试中的 React 组件。Enzyme 的 API 和 jQuery 操作 DOM 一样灵活易用。（因为其使用的是 cheerio 库来解析虚拟 DOM，而 cheerio 的目标则是做服务器端的 jQuery）Enzyme 兼容大多数断言库和测试框架，如 chai、mocha、jasmine 等

shallow 方法并将结果存储在一个变量中。然后, 我们将为在其虚拟 dom 中渲染的不同的 React 元素 (HTML 或子组件) 查询 渲染的组件
render 方法将 React 组件渲染成静态的 HTML 字符串，然后分析这段 HTML 代码的结构，返回一个对象。它跟 shallow 方法非常像，主要的不同是采用了第三方 HTML 解析库 Cheerio，它返回的是一个 Cheerio 实例对象
mount 方法用于将 React 组件加载为真实 DOM 节点。

Shallow? Mount?
shallow() 渲染函数只渲染我们专门测试的组件, 它不会渲染子元素。相反, 我们将不得不 mount() 组件, 因为子组件 Header 不可用的 jsdom, 否则。

```code
.get(index)：返回指定位置的子组件的DOM节点
.at(index)：返回指定位置的子组件
.first()：返回第一个子组件
.last()：返回最后一个子组件
.type()：返回当前组件的类型np
.text()：返回当前组件的文本内容
.html()：返回当前组件的HTML代码形式
.props()：返回根组件的所有属性
.prop(key)：返回根组件的指定属性
.state([key])：返回根组件的状态
.setState(nextState)：设置根组件的状态
.setProps(nextProps)：设置根组件的属性
.simulate('click')；方法就在这个组件上触发某种行为
.simulate(event, mock)：用来模拟事件触发，event为事件名称，mock为一个event object
.instance()：返回测试组件的实例；
.find(selector) 查找选择器下的DOM 元素，返回一个数组。
.contains(node) 确定是否包含该节点或者一些节点 ，返回true 或者 false
.is(selector) 判断改节点是否能够匹配选择器的节点 ，返回true 或者 false
.hasClass(className) 判断是否包含这个类，返回true 或者 false
```

## 配置 react+jest+enzymeem 测试

```code
npm install babel-jest babel-core@^7.0.0-bridge.0  enzyme enzyme-adapter-react-16  jest jest-stare react-test-renderer --save-dev
```

jest jest-stare ：生成 测试 报表文档
react-test-renderer: 生成 react 快照

配置 jest.config.js

切记 .babelrc

```code
"@babel/env", node_models:不能为false
```

```code
const path = require('path');
console.log(path.resolve(__dirname));
module.exports = {
    rootDir: path.resolve(__dirname),
    collectCoverage: true, // 是否收集测试时的覆盖率信息
    //collectCoverageFrom: ['<rootDir>/src/**/*.{js,jsx,mjs}'], // 哪些文件需要收集覆盖率信息
    // coverageDirectory: '<rootDir>/coverage', // 输出覆盖信息文件的目录
    coveragePathIgnorePatterns: ['<rootDir>/node_modules/', '<rootDir>//build/', '<rootDir>/src/index.jsx'], // 统计覆盖信息时需要忽略的文件
    moduleNameMapper: {
        // 主要用于与webpack的resolve.alias匹配，注意正则写法
        '^src(.*)$': '<rootDir>/src$1'
    },
    setupFiles: ['<rootDir>/test/setup.js'], // 运行测试前可运行的脚本，比如注册enzyme的兼容
    testMatch: [
        // 匹配的测试文件
        '<rootDir>/test/**/?(*.)(spec|test).js'
    ],
    testURL: 'http://localhost/', // 运行环境下的URl
    transform: {
        //引用的文件配置
        '^.+\\.(js|jsx|mjs)$': '<rootDir>/node_modules/babel-jest',
        '^.+\\.(css|less|scss)$': '<rootDir>/test/cssTransform.js',
        '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$': '<rootDir>/test/fileTransform.js'
    },
    testPathIgnorePatterns: ['/node_modules/'],

    transformIgnorePatterns: [
        // 转换时需要忽略的文件
        '[/\\\\]node_modules[/\\\\].+\\.(js|jsx|mjs)$'
    ]
};
```

> 设置 setup.js

```code
import {configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({adapter: new Adapter()});
```

> css 引用报错 cssTransform.js

````code
module.exports = {
    process() {
        return 'module.exports = {};';
    },
    getCacheKey() {
        return 'cssTransform';
    }
};

>文件引用报错 fileTransform.js
```code
const path = require('path');
module.exports = {
    process(src, filename) {
        return `module.exports = ${JSON.stringify(path.basename(filename))};`;
    }
};
````

设置快捷方式 package.json

```code
   "test-reportone": "jest --testResultsProcessor=jest-stare ",
    "test": "jest"
```
