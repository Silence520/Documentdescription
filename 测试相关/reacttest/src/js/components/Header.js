import React from 'react';
import ReactDOM from 'react-dom';
export default class Header extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            numone: 2,
            numtwo: 2,
            numres: 0
        };
    }
    componentWillMount() {
        console.log('componentWillMount');
    }

    componentWillReceiveProps() {
        console.log(componentWillReceiveProps);
    }

    HandelChange = (e, name) => {
        this.setState({
            [name]: e.target.value
        });
    };

    HandelCount = val => {
        const {numone, numtwo} = this.state;
        let numres = val == 'add' ? parseInt(numone) + parseInt(numtwo) : parseInt(numone) - parseInt(numtwo);
        this.setState({
            numres
        });
    };
    render() {
        const {numone, numtwo, numres} = this.state;
        // console.log(this.props);
        return (
            <div className="container">
                <ul>
                    <li className="libox">12</li>
                    <li className="libox">123</li>
                    <li className="libox">12</li>
                </ul>
                <h2>using from init</h2>
                <div className="HeaderBox" />
                <div>
                    Input 1
                    <input type="text" placeholder="Input 1" value={numone} onChange={e => this.HandelChange(e, 'numone')} />
                </div>
                <div>
                    Input 2 :
                    <input type="text" placeholder="Input 2" value={numtwo} onChange={e => this.HandelChange(e, 'numtwo')} />
                </div>
                <div>
                    Output :
                    <span>{numres} </span>
                </div>
                <div>
                    <button id="add" onClick={() => this.HandelCount('add')}>
                        Add
                    </button>
                    <button id="subtract" onClick={() => this.HandelCount('sub')}>
                        <span>Subtract</span>
                    </button>
                </div>
                <hr />
            </div>
        );
    }
}
