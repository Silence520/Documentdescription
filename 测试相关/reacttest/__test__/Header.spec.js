import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import Header from '../src/js/components/Header';

// it('componentWillReceiveProps', () => {
//     let wrapper = shallow(<Header role={''} />);
//     sinon.stub(Header.prototype, 'componentWillReceiveProps');
//     wrapper.setProps({
//         role: 'admin'
//     });
//     expect(Header.prototype.componentWillReceiveProps.calledOnce).toBeTruthy();
// });
//
//
describe('Snapshot', () => {
    //获取快照
    it('Header', () => {
        const renderedValue = renderer.create(<Header />).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });
});

describe('验证header组件的event正确性', () => {
    //event事件
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<Header />);
    });

    it('事件操作加', () => {
        wrapper.setState({numone: 8, numtwo: 2});
        wrapper.find('#add').simulate('click');
        expect(wrapper.state().numres).toBe(10);
    });

    it('事件操作减', () => {
        wrapper.setState({numone: 8, numtwo: 2});
        wrapper.find('#subtract').simulate('click');
        expect(wrapper.state().numres).toBe(6);
    });

    it('事件操作减和', () => {
        const mockEvent = {
            keyCode: 13, // enter 事件
            target: {
                value: 10
            }
        };
        wrapper
            .find('input')
            .at(0)
            .simulate('change', mockEvent); //simulate 触发事件不加on 并且 设置value 要与数据格式一致
        expect(wrapper.state().numone).toBe(10);
        wrapper.find('#add').simulate('click');
        expect(wrapper.state().numres).toBe(12);
    });
});

describe('验证header组件的正确性', () => {
    //静态UI的正确性
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<Header name={'hello'} />);
    });

    it('初始化验证', () => {
        expect(wrapper.contains(<h2>using from init</h2>)).toBe(true);
        expect(wrapper.state('numone')).toBe(2);
        expect(wrapper.state('numtwo')).toBe(2);
        expect(wrapper.find('button').length).toBe(2);
        expect(wrapper.find('#add').text()).toBe('Add');
        expect(wrapper.find('#add').type()).toBe('button');
        expect(wrapper.find('#subtract').html()).toEqual(`<button id="subtract"><span>Subtract</span></button>`); //不加事件
        expect(wrapper.find('span').length).toBe(2);
        expect(wrapper.name()).toEqual('div'); //组件名称
        expect(wrapper.find('#add').name()).toEqual('button');
        expect(wrapper.find('.HeaderBox').text()).toEqual('');
        expect(wrapper.find('#subtract').children().length).toBe(1);
        expect(
            wrapper
                .find('#add')
                .parent()
                .is('div')
        ).toBe(true);

        expect(
            wrapper
                .find('#subtract')
                .childAt(0)
                .type()
        ).toEqual('span');
        expect(wrapper.contains(<h2>using from init</h2>)).toEqual(true);
        expect(wrapper.contains(<h2 />)).toEqual(false); //包含
        expect(wrapper.containsAllMatchingElements([<h2>using from init</h2>, <div className="HeaderBox" />])).toEqual(true); //数组包含all
        expect(wrapper.containsAnyMatchingElements([<h2>using from init</h2>, <div>Goodbye</div>])).toEqual(true); //数组包含 any
        expect(wrapper.containsMatchingElement(<h2>using from init</h2>)).toEqual(true);
        expect(wrapper.containsMatchingElement(<button>Add</button>)).toEqual(true); //模糊匹配html元素属性 内容精确匹配
        // console.log(wrapper.debug());
        expect(wrapper.find('.libox').every('.libox')).toEqual(true);
        expect(wrapper.find('.libox').everyWhere(n => n.hasClass('libox'))).toEqual(true);
        expect(
            wrapper.find('.libox').everyWhere(n => {
                // console.log(n.debug());
                return n.type('li');
            })
        ).toEqual(true);
        expect(wrapper.exists('#add')).toEqual(true); //判断节点是否存在
        expect(wrapper.find('.libox').filter('.add').length).toBe(0);
        const complexFoo = wrapper.find('li').filterWhere(n => typeof n.type() === 'string');
        //条件过滤器;
        expect(complexFoo.length).toBe(3);
        const complexComponents = wrapper.find('ul').findWhere(n => n.type() !== 'string');
        // console.log(complexComponents.debug());
        expect(complexComponents.length).toBe(6);
        wrapper.find('li').forEach(node => {
            //循环执行判断
            expect(node.hasClass('libox')).toEqual(true);
        });
        // console.log(wrapper.find('.libox').get(1)); //get 返回子 at 返回自己包括子
        expect(
            wrapper
                .find('.libox')
                .at(1)
                .props().children
        ).toEqual('123');
        const texts = wrapper.find('.libox').map(node => node.text());
        // console.log(texts);
        expect(texts).toEqual(['12', '123', '12']); //获取所有值
        // console.log(wrapper.props().debug());
        // console.log(wrapper.prop('name'));
        // expect(wrapper.props().name).toEqual('hello'); mount
    });
});
