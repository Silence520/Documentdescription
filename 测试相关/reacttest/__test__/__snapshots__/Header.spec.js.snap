exports[`Snapshot Header 1`] = `
<div
  className="container">
  <ul>
    <li
      className="libox">
      12
    </li>
    <li
      className="libox">
      123
    </li>
    <li
      className="libox">
      12
    </li>
  </ul>
  <h2>
    using from init
  </h2>
  <div
    className="HeaderBox" />
  <div>
    Input 1
    <input
      onChange={[Function]}
      placeholder="Input 1"
      type="text"
      value={2} />
  </div>
  <div>
    Input 2 :
    <input
      onChange={[Function]}
      placeholder="Input 2"
      type="text"
      value={2} />
  </div>
  <div>
    Output :
    <span>
      0
       
    </span>
  </div>
  <div>
    <button
      id="add"
      onClick={[Function]}>
      Add
    </button>
    <button
      id="subtract"
      onClick={[Function]}>
      <span>
        Subtract
      </span>
    </button>
  </div>
  <hr />
</div>
`;
